File: Python/DP/M-198. House Robber.py
class Solution:
    @classmethod
    def rob(self, nums: list[int]) -> int:
        n = len(nums)
        dp = [0 for _ in range(n)]
        dp[0] = nums[0]
        for i in range(1, n):
            prev = dp[i - 2] if i >= 2 else 0
            dp[i] = max(prev + nums[i], dp[i - 1])

        return dp[n - 1]


testData = [
    {
        "nums": [1, 2, 3, 1],
        "ans": 4,
    },
    {
        "nums": [2, 7, 9, 3, 1],
        "ans": 12,
    },
]

for data in testData:
    print(data["ans"] == Solution.rob(data["nums"]))

File: DP/M-198. House Robber/M-198. House Robber.ts
function rob(nums: number[]): number {
  const n = nums.length;
  const dp = new Array(n).fill(0);
  dp[0] = nums[0];
  for (let i = 1; i < n; i++) {
    dp[i] = Math.max((dp[i - 2] || 0) + nums[i], dp[i - 1]);
  }
  return dp[n - 1];
}

const testData = [
  {
    nums: [1, 2, 3, 1],
    ans: 4,
  },
  {
    nums: [2, 7, 9, 3, 1],
    ans: 12,
  },
];

for (const data of testData) {
  console.log(data.ans === rob(data.nums));
}
